Modules/_winapi.c

```diff
@@ -774,12 +774,139 @@ gethandle(PyObject* obj, const char* name)
+    if (kwnames == NULL) {
+        goto error;
+    }
+    PyObject *args[] = { keys, keyfunc };
+    if (PyObject_VectorcallMethod(&_Py_ID(sort), args, 1, kwnames) == NULL) {
+        goto error;
+    }
+
+    dedupped_keys = PyList_New(0);
+    if (dedupped_keys == NULL) {
+        goto error;
+    }
+    wchar_t *prev_key_string = NULL;
+    for (Py_ssize_t i=PyList_GET_SIZE(keys)-1; i>=0; i--) {
+        PyObject *key = PyList_GET_ITEM(keys, i);
+        if (key == NULL) {
+            goto error;
+        }
+        if (i == PyList_GET_SIZE(keys)) {
+            prev_key_string = PyUnicode_AsWideCharString(key, NULL);
+            if (prev_key_string == NULL) {
+                goto error;
+            }
+            if (PyList_Insert(dedupped_keys, 0, key) < 0) {
+                goto error;
+            }
+            continue;
+        }
+        wchar_t *key_string = PyUnicode_AsWideCharString(key, NULL);
+        if (key_string == NULL) {
+            goto error;
+        }
+        if (CompareStringOrdinal(prev_key_string, -1, key_string, -1, TRUE) == CSTR_EQUAL) {
+            PyMem_Free(key_string);
+            continue;
+        }
+        if (PyList_Insert(dedupped_keys, 0, key) < 0) {
+            goto error;
+        }
+        PyMem_Free(prev_key_string);
+        prev_key_string = key_string;
+    }
+
+    result = PyDict_New();
+    if (result == NULL) {
+        goto error;
+    }
+    for (int i=0; i<PyList_GET_SIZE(dedupped_keys); i++) {
+        PyObject *key = PyList_GET_ITEM(dedupped_keys, i);
+        if (key == NULL) {
+            goto error;
+        }
+        PyObject* value = PyObject_GetItem(environment, key);
+        if (value == NULL) {
+            goto error;
+        }
+
+        if (! PyUnicode_Check(key) || ! PyUnicode_Check(value)) {
+            PyErr_SetString(PyExc_TypeError,
+                "environment can only contain strings");
+            goto error;
+        }
+        if (PyUnicode_FindChar(key, '\0', 0, PyUnicode_GET_LENGTH(key), 1) != -1 ||
+            PyUnicode_FindChar(value, '\0', 0, PyUnicode_GET_LENGTH(value), 1) != -1)
+        {
+            PyErr_SetString(PyExc_ValueError, "embedded null character");
+            goto error;
+        }
+        /* Search from index 1 because on Windows starting '=' is allowed for
+           defining hidden environment variables. */
+        if (PyUnicode_GET_LENGTH(key) == 0 ||
+            PyUnicode_FindChar(key, '=', 1, PyUnicode_GET_LENGTH(key), 1) != -1)
+        {
+            PyErr_SetString(PyExc_ValueError, "illegal environment variable name");
+            goto error;
+        }
+
+        if (PyObject_SetItem(result, key, value) < 0) {
+            goto error;
+        }
+    }
+
+    goto cleanup;
+error:
+    Py_XDECREF(result);
+    result = NULL;
+cleanup:
+    Py_XDECREF(keys);
+    Py_XDECREF(dedupped_keys);
+    Py_XDECREF(keyfunc);
+    Py_XDECREF(kwnames);
+    if (prev_key_string != NULL) {
```

Comment:
There is a risk that `prev_key_string` is uninitialized at this point. Did you consider to break up `normalize_environment` in multiple functions, in order to simplify the resource management (ref handling, memory handling)?

Reply:
Thanks for the review! I splited the function into two, it should be better to read now.

Reply:
Thanks, I'll have another look!